services/spar/src/Spar/Data.hs

```diff
@@ -281,28 +281,62 @@ getSAMLSomeUsersByIssuer issuer =
     <$$> (retry x1 . query sel $ params Quorum (Identity issuer))
   where
     sel :: PrepQuery R (Identity SAML.Issuer) (SAML.NameID, UserId)
-    sel = "SELECT sso_id, uid FROM user WHERE issuer = ? LIMIT 2000"
+    sel = "SELECT sso_id, uid FROM user_v2 WHERE issuer = ? LIMIT 2000"
 
 getSAMLUser :: (HasCallStack, MonadClient m) => SAML.UserRef -> m (Maybe UserId)
-getSAMLUser (SAML.UserRef tenant subject) =
-  runIdentity
-    <$$> (retry x1 . query1 sel $ params Quorum (tenant, subject))
-  where
-    sel :: PrepQuery R (SAML.Issuer, SAML.NameID) (Identity UserId)
-    sel = "SELECT uid FROM user WHERE issuer = ? AND sso_id = ?"
+getSAMLUser uref = do
+  mbUid <- getSAMLUserNew uref
+  case mbUid of
+    Nothing -> migrateLegacy uref
+    Just uid -> pure $ Just uid
+  where
+    getSAMLUserNew :: (HasCallStack, MonadClient m) => SAML.UserRef -> m (Maybe UserId)
+    getSAMLUserNew (SAML.UserRef tenant subject) =
+      runIdentity
+        <$$> (retry x1 . query1 sel $ params Quorum (tenant, normalizeQualifiedNameId subject))
+      where
+        sel :: PrepQuery R (SAML.Issuer, NormalizedUNameID) (Identity UserId)
+        sel = "SELECT uid FROM user_v2 WHERE issuer = ? AND normalized_uname_id = ?"
+
+    migrateLegacy :: (HasCallStack, MonadClient m) => SAML.UserRef -> m (Maybe UserId)
+    migrateLegacy uref' = do
+      mbUid <- getSAMLUserLegacy uref'
+      for mbUid $ \uid -> do
+        insertSAMLUser uref' uid
+        pure uid
+
+    getSAMLUserLegacy :: (HasCallStack, MonadClient m) => SAML.UserRef -> m (Maybe UserId)
+    getSAMLUserLegacy (SAML.UserRef tenant subject) =
+      runIdentity
+        <$$> (retry x1 . query1 sel $ params Quorum (tenant, subject))
+      where
+        sel :: PrepQuery R (SAML.Issuer, SAML.NameID) (Identity UserId)
+        sel = "SELECT uid FROM user WHERE issuer = ? AND sso_id = ?"
 
 deleteSAMLUsersByIssuer :: (HasCallStack, MonadClient m) => SAML.Issuer -> m ()
 deleteSAMLUsersByIssuer issuer = retry x5 . write del $ params Quorum (Identity issuer)
   where
     del :: PrepQuery W (Identity SAML.Issuer) ()
-    del = "DELETE FROM user WHERE issuer = ?"
-
--- | Delete a user from the saml users table.
-deleteSAMLUser :: (HasCallStack, MonadClient m) => SAML.UserRef -> m ()
-deleteSAMLUser (SAML.UserRef tenant subject) = retry x5 . write del $ params Quorum (tenant, subject)
-  where
-    del :: PrepQuery W (SAML.Issuer, SAML.NameID) ()
-    del = "DELETE FROM user WHERE issuer = ? AND sso_id = ?"
+    del = "DELETE FROM user_v2 WHERE issuer = ?"
+
+deleteSAMLUser :: (HasCallStack, MonadClient m) => UserId -> SAML.UserRef -> m ()
+deleteSAMLUser uid uref = do
+  muidUref <- getSAMLUser uref
+  for_ muidUref $ \uidUref ->
+    when (uidUref == uid) $ do
```

Comment:
The only far-fetched scenario where this should fail is when there we have to NameId A(user1), B(user2) that would collide under the equivalence. Let's assume we resolve the colllision to user1, so in effect we have A->user1 B->user1. If there is request to delete B with the intention to delete user2 this guard prevents this. I'm not sure if it can even happen, but I added this guard because it doesn't harm. The fact that call call sites of `deleteSAMLUser` have the `uid` available probably means that it can't happen.